### DISCUSSION TITLE
向量录入速度

### DISCUSSION THREAD

**Discussion by liuyichen2013** [milvus-io/milvus#18075] [Q&A and General discussion]
我看官方的性能测试报告，录入100w128维向量大概30多秒；我单机版，版本2.02测试大概需要200多秒（先创建索引、load，然后再insert，每次50000条向量），怎么会差距这么大？
代码差不多如下：
dim = 128
default_fields = [
        FieldSchema(name="count", dtype=DataType.INT64, is_primary=True),
        FieldSchema(name="float_vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
    ]
    default_schema = CollectionSchema(fields=default_fields, description="test collection")

    has = utility.has_collection(collectionName)
    print(f"Does collection {collectionName} exist in Milvus: {has}")

    #print(f"\nCreate collection...")

    collection = Collection(name=collectionName, schema=default_schema)

    if (has==False):
        default_index = {"index_type": "IVF_PQ", "params": {"nlist": 4096,"m": 4}, "metric_type": “IP”}
        print(f"\nCreate index...")
        collection.create_index(field_name="float_vector", index_params=default_index)
        print(f"\nload collection...")
        collection.load()
		
    nb = 50000
	
    vectors = [[random.random() for _ in range(dim)] for _ in range(nb)]
    #print(f"\nBegin instert.......................")
    logger.info('Begin instert......')
    collection.insert(
        [
            [i for i in range(nb)],
            vectors
        ]
    )

**Comment by yhmo** [comment]
统计耗时的时候是不是把vectors = [[random.random() for _ in range(dim)] for _ in range(nb)]这句产生随机向量的语句也算进去了？循环20次的话，应该是累加collection.insert()这个调用的耗时。

**Reply by liuyichen2013** [reply]
logger.info('Begin instert......')
collection.insert(
    [
        [i for i in range(nb)],
        vectors
    ]
)
logger.info('Insert successed, times = ' + str(times))

日志：
2022-07-05 11:03:10,348 - INFO - Begin instert......
2022-07-05 11:03:21,892 - INFO - Insert successed, times = 215
2022-07-05 11:03:22,917 - INFO - Begin instert......
2022-07-05 11:03:34,245 - INFO - Insert successed, times = 216
2022-07-05 11:03:35,319 - INFO - Begin instert......
2022-07-05 11:03:48,328 - INFO - Insert successed, times = 217
通过日志可以看出，仅insert函数要执行11-12秒的时间

**Comment by yhmo** [comment]
这是我本地测试的脚本：

```
import time
import random

from pymilvus import (
    connections,
    utility,
    FieldSchema, CollectionSchema, DataType,
    Collection,
)


connections.connect(host="localhost", port="19530")

dim = 128
collectionName = "t3"

default_fields = [
    FieldSchema(name="count", dtype=DataType.INT64, is_primary=True),
    FieldSchema(name="float_vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
]
default_schema = CollectionSchema(fields=default_fields, description="test collection")

has = utility.has_collection(collectionName)
print(f"Does collection {collectionName} exist in Milvus: {has}")

# print(f"\nCreate collection...")

collection = Collection(name=collectionName, schema=default_schema)

if (has == False):
     default_index = {"index_type": "IVF_PQ", "params": {"nlist": 4096, "m": 4}, "metric_type": "IP"}
     print(f"\nCreate index...")
     collection.create_index(field_name="float_vector", index_params=default_index)
     print(f"\nload collection...")
     collection.load()

nb = 50000

total_cost = 0.0
for i in range(20):
    vectors = [[random.random() for _ in range(dim)] for _ in range(nb)]
    data = [
            [i for i in range(nb)],
            vectors
        ]
    print(f"\nBegin instert.......................", i)
    start = time.time()
    collection.insert(data)
    end = time.time()
    total_cost = total_cost + (end-start)
    print("Insert", nb, "entities time cost", (end-start)*1000, "ms")

print("Total cost for insert", total_cost, "seconds")

utility.drop_collection(collectionName)

```

我也是用的2.0.2 standalone，结果如下：
```
Begin instert....................... 0
Insert 50000 entities time cost 2822.1020698547363 ms

Begin instert....................... 1
Insert 50000 entities time cost 2811.4120960235596 ms

Begin instert....................... 2
Insert 50000 entities time cost 3279.049873352051 ms

Begin instert....................... 3
Insert 50000 entities time cost 2720.50404548645 ms

Begin instert....................... 4
Insert 50000 entities time cost 4719.457149505615 ms

Begin instert....................... 5
Insert 50000 entities time cost 3317.3880577087402 ms

Begin instert....................... 6
Insert 50000 entities time cost 3188.7869834899902 ms

Begin instert....................... 7
Insert 50000 entities time cost 3093.2271480560303 ms

Begin instert....................... 8
Insert 50000 entities time cost 3417.1712398529053 ms

Begin instert....................... 9
Insert 50000 entities time cost 3058.04181098938 ms

Begin instert....................... 10
Insert 50000 entities time cost 3206.7577838897705 ms

Begin instert....................... 11
Insert 50000 entities time cost 2826.772928237915 ms

Begin instert....................... 12
Insert 50000 entities time cost 3486.6178035736084 ms

Begin instert....................... 13
Insert 50000 entities time cost 2773.447036743164 ms

Begin instert....................... 14
Insert 50000 entities time cost 3239.2618656158447 ms

Begin instert....................... 15
Insert 50000 entities time cost 4795.69411277771 ms

Begin instert....................... 16
Insert 50000 entities time cost 3845.8831310272217 ms

Begin instert....................... 17
Insert 50000 entities time cost 3032.080888748169 ms

Begin instert....................... 18
Insert 50000 entities time cost 3183.7708950042725 ms

Begin instert....................... 19
Insert 50000 entities time cost 2940.4850006103516 ms
Total cost for insert 65.75791192054749 seconds
```

虽然没有30多秒那么快，大约是65-70秒这样。对比这里https://milvus.io/cn/docs/v2.0.x/benchmark.md 所列举的硬件配置，可能是输在CPU上，页面里写的是至强CPU，所以......


PS：我们鼓励使用英文提问哈

**Reply by xiaofan-luan** [reply]
The lastest master improve the ingestion performance by change the compression algorithm.
Try it !

**Reply by xiaofan-luan** [reply]
From our test, the ingestion performance is increased to similar to 2.0.2
